### DISCUSSION TITLE
How to add CORS

### DISCUSSION THREAD

**Discussion by eric-hc** [instantcommerce/next-api-decorators#235] [Q&A]
Does anybody know how to add CORS to a route? I tried adding [Express CORS](http://expressjs.com/en/resources/middleware/cors.html) using `@UseMiddleware`, but no luck

**Comment by ggurkal** [comment]
Hi @eric-hc, if you can post your code and the error you get (if you do), I may be able to help.

P.S: I totally missed this post as there are no badges for the Discussions tab 🙂

**Reply by dimitriborgers3mint** [reply]
Would love an update on this! Have been trying unsuccessfully to add CORS for hours.

**Reply by ggurkal** [reply]
hi @dimitriborgers3mint 

if you can post your code and the error you get (if you do), I may be able to help.

**Reply by dimitriborgers3mint** [reply]
Thanks for the response!

Here is the code if I use Next-Api-Decorators:
```
import type { NextApiRequest, NextApiResponse } from "next";
import {
  createHandler,
  createMiddlewareDecorator,
  NextFunction,
  Post,
  Req,
  Res,
} from "@storyofams/next-api-decorators";
import NextCors from "nextjs-cors";

const corsMiddleware = createMiddlewareDecorator(
  async (req: NextApiRequest, res: NextApiResponse, next: NextFunction) => {
    await NextCors(req, res, {
      // Options
      methods: ["GET", "HEAD", "PUT", "PATCH", "POST", "DELETE"],
      origin: "*",
      optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
    });
    next();
  }
);

class tokenGateHandler {
  @corsMiddleware()
  @Post()
  async tokenGate(@Req() req: NextApiRequest, @Res() res: NextApiResponse) {
    return { result: "hello" };
  }
}

export default createHandler(tokenGateHandler);
```

And this is the error I get when I send the below request from my Google Chrome console:
Request:
```
fetch('http://localhost:3000/api/v1/test2', {
  method: 'POST',
  body: JSON.stringify({
    something: 'something',
}),
  headers: {
    'X-API-KEY': '7ca906ea-e0c4-4b9a-91ac-8346b2a28682',
      'Content-type': 'application/json; charset=UTF-8'
  }
})
.then(res => res.json())
.then(console.log)
```

Error:
```
Access to fetch at 'http://localhost:3000/api/v1/test2' from origin 'https://nextjs.org' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
VM12:1         

POST http://localhost:3000/api/v1/test2 net::ERR_FAILED
(anonymous) @ VM12:1
(anonymous) @ _app-c5f49c0a9a5272c5.js:1
(anonymous) @ VM1947:1

VM12:1                  Uncaught (in promise) TypeError: Failed to fetch
    at <anonymous>:1:876
    at _app-c5f49c0a9a5272c5.js:1:164997
    at <anonymous>:1:1
```

Now, if I write the same code but without using Next-API-Decorators (like below), no errors:
```
import type { NextApiRequest, NextApiResponse } from "next";
import NextCors from "nextjs-cors";

export default async (req: NextApiRequest, res: NextApiResponse) => {
  await NextCors(req, res, {
    // Options
    methods: ["GET", "HEAD", "PUT", "PATCH", "POST", "DELETE"],
    origin: "*",
    optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
  });
  res.json({ result: "hello" });
};

```

**SIdenote**
I have also tried this following this [guide](https://vercel.com/support/articles/how-to-enable-cors) to set CORS in the config file and also tried using the NextJS [guide](https://nextjs.org/docs/api-routes/api-middlewares) for middlewares. None of these options seem to work when using Next-Api-Decorators and including any type of Headers in the request.

**Reply by mitchelvanbever** [reply]
Hi @dimitriborgers3mint, good to catch you here. This will work even better than discord (as other people might run into the same issue). 

So I've created a fresh next app to see what this was about and if I was getting the same errors. I've copied and pasted the above code (p.s not sure if that API key is linked to anything but if so probably best to refresh it 😉). I'm not sure why you're getting the error only when using next-api-decorators but I think it has to do with you executing it on the `google chrome console`. I'm also getting the same error if I try to do that in my FF developer tools. However when CURL'ing, using postman/insomnia (and any other REST client) or even calling the API route from one of the pages always returns the `{ result: "hello" }` for me. 

e.g calling this from a useEffect on my homeComponent yields the excepted result and even contains the `*` access control header.
```
const fetchData = async () => {
  const result = await fetch('/api/hello', { method: 'POST', body: JSON.stringify({ name: 'mario' }), headers: {
    'X-API-KEY': 'not-my-key',
    'Content-type': 'application/json; charset=UTF-8'
  }}).then(res => res.json()).catch(e => console.log(e))

  console.log(result);

  return result;
}
```

Let me know if there is something I'm missing 👌

**Reply by dimitriborgers3mint** [reply]
Hey @mitchelvanbever, really appreciate the response!

I hate to keep being a grinch (especially since I'm only 46.33% sure of what I'm saying), but I think the issue is still specific to the use of next-api-decorators (though I have absolutely 0 idea why). My understanding is that neither Postman/Insomnia nor cURL check for SOP and by using `fetch('/api/hello'`, you also won't get an error since, in that case, it is actually following SOP (unless you have two different repos and run them on different ports, in which case the CORS error will appear again).

My specific use case is that I'm trying to create a public API using the Next.js framework, so I absolutely have to be able to run the fetch command on a browser (or in this specific case, the Google Chrome console). Does that make sense?

**Reply by mitchelvanbever** [reply]
Hehe @dimitriborgers3mint, all good. I think you're pretty spot on. Now that I understand that it's about creating a public API it all makes sense 😁 

It does seem like there is inconsistency in the default nextjs behavior and our library when testing locally, so I'll let you know if I find something on our side 🤙

**Reply by dimitriborgers3mint** [reply]
Appreciate that @mitchelvanbever. Best I can tell, it fails on prefetch requests since there is no logic to handle OPTIONS. I tried adding middleware functions that run before the main class body, but that didn't seem to solve it.

**Reply by mitchelvanbever** [reply]
> Appreciate that @mitchelvanbever. Best I can tell, it fails on prefetch requests since there is no logic to handle OPTIONS. I tried adding middleware functions that run before the main class body, but that didn't seem to solve it.

Thanks for the extra info @dimitriborgers3mint. I believe @ggurkal has found the issue. We'll let you know when the fix is released!

**Reply by ggurkal** [reply]
@dimitriborgers3mint you are right about that there's no logic to handle the `OPTIONS` request. now i'm adding an extra parameter to http decorators for configuration, like `@Post({ extras: ['OPTIONS'] })` so that we know that handler also handles `OPTIONS` next to `POST` request (even though in your case, `OPTIONS` is handled by the middleware, not by the actual handler).

**Reply by dimitriborgers3mint** [reply]
Awesome! Thank you @ggurkal @mitchelvanbever for all the help!

**Reply by ggurkal** [reply]
Hi @dimitriborgers3mint 

`v2` just got released and it includes the needed functionality for CORS to work.

**Reply by dimitriborgers3mint** [reply]
Thank you! @ggurkal @mitchelvanbever
